src/blocking/client.rs

```diff
@@ -41,74 +39,70 @@ use crate::senum::{
     AdditionalType, AlbumType, Country, IncludeExternal, RepeatState, SearchType, TimeRange, Type,
 };
 
-/// Describes API errors
-#[derive(Debug, Deserialize)]
+/// Possible errors returned from the `rspotify` client.
+#[derive(Debug, Error)]
+pub enum ClientError {
+    #[error("request unauthorized")]
+    UnauthorizedError,
+    #[error("exceeded request limit")]
+    RateLimitedError(Option<usize>),
+    #[error("spotify error: {0}")]
+    ApiError(#[from] ApiError),
+    #[error("json parse error: {0}")]
+    ParseJSONError(#[from] serde_json::Error),
+    #[error("request error: {0}")]
+    RequestError(#[from] reqwest::Error),
+    #[error("status code: {0}")]
+    StatusCodeError(StatusCode),
+}
+
+impl From<Response> for ClientError {
+    fn from(response: Response) -> Self {
+        match response.status() {
+            StatusCode::UNAUTHORIZED => Self::UnauthorizedError,
+            StatusCode::TOO_MANY_REQUESTS => Self::RateLimitedError(
+                response
+                    .headers()
+                    .get(reqwest::header::RETRY_AFTER)
```

Comment:
This was rewritten because indexing directly can panic (though it is unlikely). Doing a match on `response.headers().get(...).and_then(|header| header.to_str().ok())` was suddenly a lot going on so I used another `and_then` instead.

Reply:
I agree, It's better this way IMO

Reply:
Yeah, it's more robust than before :0